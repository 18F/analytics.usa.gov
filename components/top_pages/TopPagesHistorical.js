import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../js/lib/renderblock";
import { exceptions, titleExceptions } from "../../js/lib/exceptions";
import barChart from "../../js/lib/barchart";
import formatters from "../../js/lib/formatters";

function TopPagesHistorical({ dataHrefBase, reportFileName }) {
  const dataURL = `${dataHrefBase}/${reportFileName}`;
  const ref = useRef(null);

  useEffect(() => {
    const initHistoricalChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => d.data)
            .on("render", (selection) => {
              // turn the labels into links
              selection
                .selectAll(".label")
                .each(function (d) {
                  d.text = this.innerText;
                })
                .html("")
                .append("a")
                .attr("target", "_blank")
                .attr("rel", "noopener")
                .attr(
                  "href",
                  (d) => exceptions[d.domain] || `http://${d.domain}`,
                )
                .text((d) => titleExceptions[d.domain] || d.domain);
            })
            .render(
              barChart()
                .label((d) => d.domain)
                .value((d) => +d.visits)
                .scale((values) =>
                  d3.scale
                    .linear()
                    .domain([0, 1, d3.max(values)])
                    .rangeRound([0, 1, 100]),
                )
                .format(formatters.addCommas),
            ),
        );
      return result;
    };
    initHistoricalChart().catch(console.error);
  });

  return (
    <figure className="bar-chart__top-pages" data-source={dataURL} ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopPagesHistorical.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  reportFileName: PropTypes.string.isRequired,
};

export default TopPagesHistorical;
