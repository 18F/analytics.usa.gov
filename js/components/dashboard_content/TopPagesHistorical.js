import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import DataLoader from "../../lib/data_loader";
import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import { exceptions, titleExceptions } from "../../lib/exceptions";
import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the top pages report from the passed data URL and report file name
 * creates a visualization for the count of users visiting sites for the current
 * agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {string} props.reportFileName the file name of the report to use as a
 * data source for the data visualization.
 * @param {number} props.numberOfListingsToDisplay the count of top page
 * listings to display in the bar chart.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopPagesHistorical({
  dataHrefBase,
  reportFileName,
  numberOfListingsToDisplay,
}) {
  const dataURL = `${dataHrefBase}/${reportFileName}`;
  const ref = useRef(null);
  const [topPagesData, setTopPagesData] = useState(null);

  useEffect(() => {
    const initTopPagessChart = async () => {
      if (!topPagesData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setTopPagesData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(topPagesData)
          .setTransformer((d) => d.data.slice(0, numberOfListingsToDisplay))
          .setRenderer((selection) => {
            const barchartRenderer = barChart()
              .label((d) => d.domain)
              .value((d) => +d.visits)
              .scale((values) =>
                d3.scale
                  .linear()
                  .domain([0, 1, d3.max(values)])
                  .rangeRound([0, 1, 100]),
              )
              .format(formatters.addCommas);
            barchartRenderer(selection);

            selection
              .selectAll(".label")
              .each(function (d) {
                d.text = this.innerText;
              })
              .html("")
              .append("a")
              .attr("target", "_blank")
              .attr("rel", "noopener")
              .attr("href", (d) => exceptions[d.domain] || `http://${d.domain}`)
              .text((d) => titleExceptions[d.domain] || d.domain);
          })
          .build();
      }
    };
    initTopPagessChart().catch(console.error);
  }, [topPagesData]);

  return (
    <figure className="top-pages__bar-chart" data-source={dataURL} ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopPagesHistorical.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  reportFileName: PropTypes.string.isRequired,
  numberOfListingsToDisplay: PropTypes.number.isRequired,
};

export default TopPagesHistorical;
