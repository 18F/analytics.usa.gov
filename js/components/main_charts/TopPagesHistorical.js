import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import { exceptions, titleExceptions } from "../../lib/exceptions";
import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the top pages report from the passed data URL and report file name
 * creates a visualization for the count of users visiting sites for the current
 * agency.
 *
 * @param {String} dataHrefBase the URL of the base location of the data to be
 * downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {String} reportFileName the file name of the report to use as a data
 * source for the data visualization.
 */
function TopPagesHistorical({ dataHrefBase, reportFileName }) {
  const dataURL = `${dataHrefBase}/${reportFileName}`;
  const ref = useRef(null);

  useEffect(() => {
    const initHistoricalChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => d.data)
            .on("render", (selection) => {
              // turn the labels into links
              selection
                .selectAll(".label")
                .each(function (d) {
                  d.text = this.innerText;
                })
                .html("")
                .append("a")
                .attr("target", "_blank")
                .attr("rel", "noopener")
                .attr(
                  "href",
                  (d) => exceptions[d.domain] || `http://${d.domain}`,
                )
                .text((d) => titleExceptions[d.domain] || d.domain);
            })
            .render(
              barChart()
                .label((d) => d.domain)
                .value((d) => +d.visits)
                .scale((values) =>
                  d3.scale
                    .linear()
                    .domain([0, 1, d3.max(values)])
                    .rangeRound([0, 1, 100]),
                )
                .format(formatters.addCommas),
            ),
        );
      return result;
    };
    initHistoricalChart().catch(console.error);
  });

  return (
    <figure className="bar-chart__top-pages" data-source={dataURL} ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopPagesHistorical.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  reportFileName: PropTypes.string.isRequired,
};

export default TopPagesHistorical;
