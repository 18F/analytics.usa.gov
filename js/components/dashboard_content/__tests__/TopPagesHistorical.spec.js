import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import TopPagesHistorical from "../TopPagesHistorical";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("TopPagesHistorical", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopPagesHistorical
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={10}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "top-pages-30-days",
        query: {},
        meta: {
          name: "Top Pages (30 Days)",
          description:
            "Last 30 days' top 20 pages, measured by page views, for all sites.",
        },
        data: [
          {
            domain: "tools.usps.com",
            page: "tools.usps.com/go/TrackConfirmAction",
            page_title: "USPS.com® - USPS Tracking® Results",
            pageviews: "175651763",
            visits: "126859658",
            active_visitors: "41122748",
            pageviews_per_session: "1.3846148237290692",
            avg_session_duration: "171.74193047494396",
            bounce_rate: "0.516912713102222",
          },
          {
            domain: "tools.usps.com",
            page: "tools.usps.com/go/TrackConfirmAction_input",
            page_title: "USPS.com® - USPS Tracking® Results",
            pageviews: "172077548",
            visits: "137854682",
            active_visitors: "53726725",
            pageviews_per_session: "1.2482531931704721",
            avg_session_duration: "143.76776746692858",
            bounce_rate: "0.71359610404817442",
          },
          {
            domain: "www.time.gov",
            page: "www.time.gov/",
            page_title: "National Institute of Standards and Technology | NIST",
            pageviews: "84090552",
            visits: "1266571",
            active_visitors: "636538",
            pageviews_per_session: "66.392291470434742",
            avg_session_duration: "5480.046422452051",
            bounce_rate: "0.32670888564478423",
          },
          {
            domain: "time.gov",
            page: "time.gov/",
            page_title: "National Institute of Standards and Technology | NIST",
            pageviews: "72639584",
            visits: "670903",
            active_visitors: "288571",
            pageviews_per_session: "108.27136560724874",
            avg_session_duration: "8884.0348247349248",
            bounce_rate: "0.32556867386194427",
          },
          {
            domain: "informeddelivery.usps.com",
            page: "informeddelivery.usps.com/box/pages/secure/DashboardAction_input.action",
            page_title: "Informed Delivery",
            pageviews: "69452793",
            visits: "38515219",
            active_visitors: "7172331",
            pageviews_per_session: "1.8032558246650499",
            avg_session_duration: "139.83037450543193",
            bounce_rate: "0.065903896327319342",
          },
          {
            domain: "reg.usps.com",
            page: "reg.usps.com/entreg/LoginAction_input",
            page_title: "USPS.com® - Sign In",
            pageviews: "48787297",
            visits: "42294001",
            active_visitors: "19010794",
            pageviews_per_session: "1.1535275889363128",
            avg_session_duration: "65.027813792656616",
            bounce_rate: "0.25750216916105906",
          },
          {
            domain: "tools.usps.com",
            page: "tools.usps.com/go/TrackConfirmAction.action",
            page_title: "USPS.com® - USPS Tracking® Results",
            pageviews: "39112726",
            visits: "31254428",
            active_visitors: "13242495",
            pageviews_per_session: "1.2514298773920931",
            avg_session_duration: "137.3760588116821",
            bounce_rate: "0.67310251206645022",
          },
          {
            domain: "www.usps.com",
            page: "www.usps.com/",
            page_title: "Welcome | USPS",
            pageviews: "38337381",
            visits: "30140169",
            active_visitors: "16102673",
            pageviews_per_session: "1.27196967608244",
            avg_session_duration: "75.4410073226582",
            bounce_rate: "0.1505468665421219",
          },
          {
            domain: "www.usps.com",
            page: "www.usps.com/manage/informed-delivery.htm",
            page_title:
              "Informed Delivery - Mail & Package Notifications | USPS",
            pageviews: "35945379",
            visits: "32447672",
            active_visitors: "8288147",
            pageviews_per_session: "1.107795314252437",
            avg_session_duration: "45.984241744616888",
            bounce_rate: "0.072226013625877383",
          },
          {
            domain: "reg.usps.com",
            page: "reg.usps.com/portal/login",
            page_title: "USPS.com® - Sign In",
            pageviews: "30299198",
            visits: "27517507",
            active_visitors: "4890637",
            pageviews_per_session: "1.1010880091717612",
            avg_session_duration: "32.001429442707845",
            bounce_rate: "0.012020856395166903",
          },
        ],
        totals: {
          visits: 571051212,
        },
        taken_at: "2024-01-05T15:31:52.352Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopPagesHistorical
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={10}
        />,
      );
      await waitFor(() => screen.getByRole("link", { name: "www.time.gov" }));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopPagesHistorical
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={10}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
