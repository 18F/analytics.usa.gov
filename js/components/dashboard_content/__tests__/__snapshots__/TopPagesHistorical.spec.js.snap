// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopPagesHistorical when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <figure
    class="top-pages__bar-chart"
    data-source="http://www.example.com/data//foobar.json"
  >
    <div
      class="data bar-chart"
    >
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://tools.usps.com"
            rel="noopener"
            target="_blank"
          >
            tools.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          126,859,658
        </div>
        <div
          class="bar"
          style="width: 92%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://tools.usps.com"
            rel="noopener"
            target="_blank"
          >
            tools.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          137,854,682
        </div>
        <div
          class="bar"
          style="width: 100%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://www.time.gov"
            rel="noopener"
            target="_blank"
          >
            www.time.gov
          </a>
        </div>
        <div
          class="value"
        >
          1,266,571
        </div>
        <div
          class="bar"
          style="width: 2%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://time.gov"
            rel="noopener"
            target="_blank"
          >
            time.gov
          </a>
        </div>
        <div
          class="value"
        >
          670,903
        </div>
        <div
          class="bar"
          style="width: 1%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://informeddelivery.usps.com"
            rel="noopener"
            target="_blank"
          >
            informeddelivery.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          38,515,219
        </div>
        <div
          class="bar"
          style="width: 29%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://reg.usps.com"
            rel="noopener"
            target="_blank"
          >
            reg.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          42,294,001
        </div>
        <div
          class="bar"
          style="width: 31%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://tools.usps.com"
            rel="noopener"
            target="_blank"
          >
            tools.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          31,254,428
        </div>
        <div
          class="bar"
          style="width: 23%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://www.usps.com"
            rel="noopener"
            target="_blank"
          >
            www.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          30,140,169
        </div>
        <div
          class="bar"
          style="width: 23%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://www.usps.com"
            rel="noopener"
            target="_blank"
          >
            www.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          32,447,672
        </div>
        <div
          class="bar"
          style="width: 24%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="http://reg.usps.com"
            rel="noopener"
            target="_blank"
          >
            reg.usps.com
          </a>
        </div>
        <div
          class="value"
        >
          27,517,507
        </div>
        <div
          class="bar"
          style="width: 21%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`TopPagesHistorical when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    class="top-pages__bar-chart"
    data-source="http://www.example.com/data//foobar.json"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;

exports[`TopPagesHistorical when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    class="top-pages__bar-chart"
    data-source="http://www.example.com/data//foobar.json"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;
